substrate/primitives/virtualization/src/tests.rs

```diff
@@ -0,0 +1,290 @@
+fn exit_works(program: &[u8]) {
+	let instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute_and_destroy("do_exit", syscall_handler, &mut state);
+	assert_eq!(ret, Err(ExecError::Trap));
+	assert!(state.exit);
+	assert_eq!(state.user.counter, 0);
+}
+
+/// Setting exit to true prevents the program from even launching.
+fn exit_prevents_program_launch(program: &[u8]) {
+	let instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: true,
+		user: State { counter: 7, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute_and_destroy("add_99", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert!(state.exit);
+	assert_eq!(state.user.counter, 7);
+	assert_eq!(state.gas_left, GAS_MAX);
+}
+
+/// Increment the counter in an endless loop until we run out of gas.
+fn run_out_of_gas_works(program: &[u8]) {
+	let instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: 100_000,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute_and_destroy("increment_forever", syscall_handler, &mut state);
+	assert_eq!(ret, Err(ExecError::OutOfGas));
+	assert!(!state.exit);
+	assert_eq!(state.user.counter, 6_666);
+	assert_eq!(state.gas_left, 0);
+}
+
+/// Call same function with different gas limits and make sure they consume the same amount of gas.
+fn gas_consumption_works(program: &[u8]) {
+	let gas_limit_0 = GAS_MAX;
+	let gas_limit_1 = gas_limit_0 / 2;
+
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: gas_limit_0,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("counter", syscall_handler, &mut state);
+	let gas_consumed = gas_limit_0 - state.gas_left;
+	assert_eq!(ret, Ok(()));
+
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: gas_limit_1,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("counter", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(gas_consumed, gas_limit_1 - state.gas_left);
+}
+
+/// Make sure that globals are reset for a new instance.
+fn memory_reset_on_instantiate(program: &[u8]) {
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("offset", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(state.user.counter, 3);
```

Comment:
should we call again on the same instance here to make sure the global has changed

Reply:
ah nvm you do it on the next test 🙈